File: apps/api/src/routes/ai-assistant.ts
import { Router } from "express";

import { AiAssistantMessageRequestSchema } from "@silo/engine/contracts/dto/ai-assistant";
import { authMiddleware } from "../middleware/auth.js";
import { requirePermission } from "../middleware/permissions.js";
import {
  AssistantThreadNotFoundError,
  createAssistantThread,
  deleteAssistantMessage,
  deleteAssistantThread,
  getAssistantExamplesResponse,
  getAssistantThreadDetails,
  listAssistantThreads,
  sendAssistantMessage,
  sendAssistantMessageStream,
} from "../services/ai-assistant-thread-service.js";
import { getAssistantRuntimeStatus } from "../services/ai-assistant-service.js";

const router = Router();

router.use(authMiddleware);

router.get(
  "/status",
  requirePermission("reports", "view"),
  async (_req, res) => {
    try {
      const data = await getAssistantRuntimeStatus();
      res.json({ success: true, data });
    } catch (err) {
      console.error("❌ [API_AI_ASSISTANT/STATUS] GET:", err);
      res.status(500).json({ success: false, error: "Erro interno" });
    }
  },
);

router.get(
  "/examples",
  requirePermission("reports", "view"),
  async (_req, res) => {
    try {
      const data = getAssistantExamplesResponse();
      res.json({ success: true, data });
    } catch (err) {
      console.error("❌ [API_AI_ASSISTANT/EXAMPLES] GET:", err);
      res.status(500).json({ success: false, error: "Erro interno" });
    }
  },
);

router.get(
  "/threads",
  requirePermission("reports", "view"),
  async (req, res) => {
    try {
      const user = req.user!;
      const data = await listAssistantThreads(user.id);
      res.json({ success: true, data });
    } catch (err) {
      console.error("❌ [API_AI_ASSISTANT/THREADS] GET:", err);
      res.status(500).json({ success: false, error: "Erro interno" });
    }
  },
);

router.post(
  "/threads",
  requirePermission("reports", "view"),
  async (req, res) => {
    try {
      const user = req.user!;
      const data = await createAssistantThread(user.id);
      res.status(201).json({ success: true, data, message: "Conversa criada com sucesso." });
    } catch (err) {
      console.error("❌ [API_AI_ASSISTANT/THREADS] POST:", err);
      res.status(500).json({ success: false, error: "Erro interno" });
    }
  },
);

router.get(
  "/threads/:threadId",
  requirePermission("reports", "view"),
  async (req, res) => {
    try {
      const user = req.user!;
      const threadId =
        typeof req.params.threadId === "string" ? req.params.threadId : null;

      if (!threadId) {
        res.status(400).json({ success: false, error: "Identificador da conversa inválido." });
        return;
      }

      const thread = await getAssistantThreadDetails(user.id, threadId);

      if (!thread) {
        res.status(404).json({ success: false, error: "Conversa não encontrada." });
        return;
      }

      res.json({ success: true, data: thread });
    } catch (err) {
      console.error("❌ [API_AI_ASSISTANT/THREADS/:ID] GET:", err);
      res.status(500).json({ success: false, error: "Erro interno" });
    }
  },
);

router.post(
  "/messages",
  requirePermission("reports", "view"),
  async (req, res) => {
    try {
      const user = req.user!;
      const parsedBody = AiAssistantMessageRequestSchema.safeParse(req.body);
      if (!parsedBody.success) {
        res.status(400).json({
          success: false,
          error: "Mensagem inválida.",
          field: "content",
        });
        return;
      }

      const response = await sendAssistantMessage(user, parsedBody.data);

      res.json({ success: true, data: response });
    } catch (err) {
      if (err instanceof AssistantThreadNotFoundError) {
        res.status(404).json({ success: false, error: err.message });
        return;
      }
      console.error("❌ [API_AI_ASSISTANT/MESSAGES] POST:", err);
      res.status(500).json({ success: false, error: "Erro interno" });
    }
  },
);

// Streaming: POST /api/ai-assistant/messages/stream
// Retorna Server-Sent Events com o pensamento do modelo em tempo real
router.post(
  "/messages/stream",
  requirePermission("reports", "view"),
  async (req, res) => {
    try {
      const user = req.user!;
      const parsedBody = AiAssistantMessageRequestSchema.safeParse(req.body);
      if (!parsedBody.success) {
        res.status(400).json({
          success: false,
          error: "Mensagem inválida.",
          field: "content",
        });
        return;
      }

      // Configura headers SSE
      res.writeHead(200, {
        "Content-Type": "text/event-stream",
        "Cache-Control": "no-cache",
        Connection: "keep-alive",
        "X-Accel-Buffering": "no",
      });

      const sendEvent = (event: string, data: unknown) => {
        res.write(`event: ${event}\ndata: ${JSON.stringify(data)}\n\n`);
      };

      // Envia evento inicial imediatamente para manter a conexão viva
      // enquanto a classificação de escopo e coleta de dados são processadas
      sendEvent("connected", { status: "processing" });

      // Heartbeat a cada 5s para evitar timeout do proxy durante o processamento inicial
      const heartbeatInterval = setInterval(() => {
        res.write(": heartbeat\n\n");
      }, 5_000);

      try {
        await sendAssistantMessageStream(user, parsedBody.data, sendEvent);
      } finally {
        clearInterval(heartbeatInterval);
      }

      res.end();
    } catch (err) {
      if (err instanceof AssistantThreadNotFoundError) {
        if (!res.headersSent) {
          res.status(404).json({ success: false, error: err.message });
        }
        return;
      }
      console.error("❌ [API_AI_ASSISTANT/MESSAGES/STREAM] POST:", err);
      if (!res.headersSent) {
        res.status(500).json({ success: false, error: "Erro interno" });
      } else {
        res.end();
      }
    }
  },
);

// DELETE /api/ai-assistant/threads/:threadId/messages/:messageId — exclui mensagem e seus artefatos
router.delete(
  "/threads/:threadId/messages/:messageId",
  requirePermission("reports", "view"),
  async (req, res) => {
    try {
      const user = req.user!;
      const threadId = typeof req.params.threadId === "string" ? req.params.threadId : null;
      const messageId = typeof req.params.messageId === "string" ? req.params.messageId : null;

      if (!threadId || !messageId) {
        res.status(400).json({ success: false, error: "Parâmetros inválidos." });
        return;
      }

      await deleteAssistantMessage(user.id, threadId, messageId);
      res.json({ success: true, message: "Mensagem excluída com sucesso." });
    } catch (err) {
      if (err instanceof AssistantThreadNotFoundError) {
        res.status(404).json({ success: false, error: err.message });
        return;
      }
      console.error("❌ [API_AI_ASSISTANT/THREADS/:ID/MESSAGES/:ID] DELETE:", err);
      res.status(500).json({ success: false, error: "Erro interno" });
    }
  },
);

// DELETE /api/ai-assistant/threads/:threadId — exclui thread inteiro e todos os artefatos
router.delete(
  "/threads/:threadId",
  requirePermission("reports", "view"),
  async (req, res) => {
    try {
      const user = req.user!;
      const threadId = typeof req.params.threadId === "string" ? req.params.threadId : null;

      if (!threadId) {
        res.status(400).json({ success: false, error: "Identificador da conversa inválido." });
        return;
      }

      await deleteAssistantThread(user.id, threadId);
      res.json({ success: true, message: "Conversa excluída com sucesso." });
    } catch (err) {
      if (err instanceof AssistantThreadNotFoundError) {
        res.status(404).json({ success: false, error: err.message });
        return;
      }
      console.error("❌ [API_AI_ASSISTANT/THREADS/:ID] DELETE:", err);
      res.status(500).json({ success: false, error: "Erro interno" });
    }
  },
);

export default router;

File: apps/backend/src/silo/api/routers/ai_assistant.py
from __future__ import annotations

import asyncio
import json
import logging
from typing import Annotated
from uuid import uuid4

from fastapi import APIRouter, Depends, Request
from fastapi.responses import StreamingResponse
from pydantic import ValidationError
from sqlalchemy.engine import Connection

from silo.ai.assistant_contracts import AiAssistantMessageRequestDto
from silo.ai.assistant_service import (
    AssistantMessageConflictError,
    AssistantThreadNotFoundError,
    create_assistant_thread,
    delete_assistant_message,
    delete_assistant_thread,
    get_assistant_examples,
    get_assistant_runtime_status,
    get_assistant_thread_details,
    list_assistant_threads,
    send_assistant_message,
    stream_assistant_message,
)
from silo.api.dependencies import CurrentUser, get_db, get_snapshot_db, require_permission
from silo.api.responses import build_success_payload, json_error_response
from silo.db.serialization import serialize_legacy_value

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/ai-assistant", tags=["ai-assistant"])

AssistantUser = Annotated[CurrentUser, Depends(require_permission("reports", "view"))]
AssistantReadDb = Annotated[Connection, Depends(get_snapshot_db)]
AssistantWriteDb = Annotated[Connection, Depends(get_db)]


def _request_request_id(request: Request) -> str:
    request_id = getattr(request.state, "request_id", None)
    if isinstance(request_id, str) and request_id:
        return request_id
    return str(uuid4())


def _serialize_response(data: object) -> object:
    return serialize_legacy_value(data)


def _stream_event(event_name: str, data: object) -> str:
    payload = json.dumps(
        _serialize_response(data), ensure_ascii=False, separators=(",", ":"), default=str
    )
    return f"event: {event_name}\ndata: {payload}\n\n"


def _heartbeat_event() -> str:
    return ": heartbeat\n\n"


def _invalid_request_response(field: str | None = None):
    return json_error_response(400, "Dados inválidos.", field=field)


@router.get("/status")
async def get_status(current_user: AssistantUser):
    del current_user
    status = await get_assistant_runtime_status()
    return build_success_payload(status.model_dump(mode="json"))


@router.get("/examples")
async def get_examples(current_user: AssistantUser):
    del current_user
    examples = get_assistant_examples()
    return build_success_payload(examples.model_dump(mode="json"))


@router.get("/threads")
async def get_threads(current_user: AssistantUser, db: AssistantReadDb):
    threads = list_assistant_threads(db, current_user.id)
    return build_success_payload(threads.model_dump(mode="json"))


@router.post("/threads")
async def post_thread(request: Request, current_user: AssistantUser, db: AssistantWriteDb):
    try:
        try:
            body = await request.json()
        except Exception:
            body = {}
        title = None
        if isinstance(body, dict):
            raw_title = body.get("title")
            if isinstance(raw_title, str):
                cleaned_title = raw_title.strip()
                title = cleaned_title or None
        thread = create_assistant_thread(db, current_user.id, title=title)
        return build_success_payload(thread.model_dump(mode="json"))
    except Exception:
        logger.exception(
            "Falha ao criar thread do assistente",
            extra={"request_id": _request_request_id(request)},
        )
        return json_error_response(500, "Erro interno do servidor")


@router.get("/threads/{thread_id}")
async def get_thread(thread_id: str, current_user: AssistantUser, db: AssistantReadDb):
    thread = get_assistant_thread_details(db, current_user.id, thread_id)
    if thread is None:
        return json_error_response(404, "Conversa não encontrada.")
    return build_success_payload(thread.model_dump(mode="json"))


@router.post("/messages")
async def post_message(request: Request, current_user: AssistantUser, db: AssistantWriteDb):
    try:
        body = await request.json()
        request_payload = AiAssistantMessageRequestDto.model_validate(body)
        response = await send_assistant_message(
            db,
            current_user,
            request_payload,
            request_id=_request_request_id(request),
        )
        return build_success_payload(response.model_dump(mode="json"))
    except ValidationError:
        return _invalid_request_response("content")
    except AssistantThreadNotFoundError as error:
        return json_error_response(404, str(error))
    except AssistantMessageConflictError as error:
        return json_error_response(409, str(error))
    except Exception:
        logger.exception(
            "Falha ao processar mensagem do assistente",
            extra={"request_id": _request_request_id(request)},
        )
        return json_error_response(500, "Erro interno do servidor")


@router.post("/messages/stream")
async def post_message_stream(request: Request, current_user: AssistantUser, db: AssistantWriteDb):
    try:
        body = await request.json()
        request_payload = AiAssistantMessageRequestDto.model_validate(body)
    except ValidationError:
        return _invalid_request_response("content")
    except Exception:
        logger.exception(
            "Falha ao ler corpo da mensagem do assistente",
            extra={"request_id": _request_request_id(request)},
        )
        return json_error_response(500, "Erro interno do servidor")

    async def event_stream():
        yield _stream_event("connected", {"status": "processing"})
        service_stream = stream_assistant_message(
            db,
            current_user,
            request_payload,
            request_id=_request_request_id(request),
        )
        pending_event = asyncio.create_task(service_stream.__anext__())
        try:
            while True:
                try:
                    event = await asyncio.wait_for(asyncio.shield(pending_event), timeout=5.0)
                except asyncio.TimeoutError:
                    if await request.is_disconnected():
                        pending_event.cancel()
                        break
                    yield _heartbeat_event()
                    continue
                except StopAsyncIteration:
                    break

                yield _stream_event(event.event, event.data)
                pending_event = asyncio.create_task(service_stream.__anext__())
        except AssistantThreadNotFoundError as error:
            yield _stream_event("error", {"content": str(error)})
        except AssistantMessageConflictError as error:
            yield _stream_event("error", {"content": str(error)})
        except asyncio.CancelledError:
            pending_event.cancel()
            await asyncio.gather(pending_event, return_exceptions=True)
            await service_stream.aclose()
            raise
        except Exception:
            logger.exception(
                "Falha ao transmitir mensagem do assistente",
                extra={"request_id": _request_request_id(request)},
            )
            yield _stream_event("error", {"content": "Erro interno do servidor"})
        finally:
            if not pending_event.done():
                pending_event.cancel()
                await asyncio.gather(pending_event, return_exceptions=True)
            await service_stream.aclose()

    return StreamingResponse(
        event_stream(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )


@router.delete("/threads/{thread_id}/messages/{message_id}")
async def delete_message(
    request: Request,
    thread_id: str,
    message_id: str,
    current_user: AssistantUser,
    db: AssistantWriteDb,
):
    try:
        delete_assistant_message(db, current_user.id, thread_id, message_id)
        return build_success_payload({"deleted": True})
    except AssistantThreadNotFoundError as error:
        return json_error_response(404, str(error))
    except Exception:
        logger.exception(
            "Falha ao excluir mensagem do assistente",
            extra={"request_id": _request_request_id(request)},
        )
        return json_error_response(500, "Erro interno do servidor")


@router.delete("/threads/{thread_id}")
async def delete_thread(
    request: Request,
    thread_id: str,
    current_user: AssistantUser,
    db: AssistantWriteDb,
):
    try:
        delete_assistant_thread(db, current_user.id, thread_id)
        return build_success_payload({"deleted": True})
    except AssistantThreadNotFoundError as error:
        return json_error_response(404, str(error))
    except Exception:
        logger.exception(
            "Falha ao excluir thread do assistente",
            extra={"request_id": _request_request_id(request)},
        )
        return json_error_response(500, "Erro interno do servidor")
